### ISSUE TITLE
UI features Delete/disable not working as they are supposed to.

### ISSUE THREAD

**Issue by ShobhitGoel95** [ConfigHubPub/ConfigHubPlatform#94]
Hey,
While using ConfigHub I found that some of the features of UI named as Delete/Disable are not working properly. 
Explanation: When am deleting/disabling any context from UI am still able to get the data through pull api request. However, it should not be the case.

**Comment by edoarsla** [comment]
Hi. A context element cannot be deleted if there are properties or files referencing it.
When you go to the context element page, it will list all properties/files that still use it.

**Comment by ShobhitGoel95** [comment]
Hi edo ,
Thanks for the response.
But if an element is not referencing any files/properties then it can be deleted. Right?
And What about disabling the context for some reason? In that case also we should not expect any output while doing PULL API?

**Comment by edoarsla** [comment]
If the element is not referencing anything, it can be deleted.
Could you explain "disabling the context" please.

**Comment by ShobhitGoel95** [comment]
As you mentioned that a context can be deleted But in this case also am getting the result not fr the context I requested for but for the default context (*;*;*;*).
Same for the case of Disable.
Example:
Suppose I have a context like this::  Product1;Dev;*;* 
and when I am requesting the data for this context using pull API am getting the correct result.
But if I remove the context from UI or disable it (Using UI only) then in this I am expecting an error to come up.The error may be anything like 404 or anything.
But what am getting is the result for default context i.e., *;*;*;* that too with the context I have requested data for and in case the value against the default context is not there am getting the result. 

![error1](https://user-images.githubusercontent.com/46491526/59689074-9c34c180-91fc-11e9-8579-acbb00653b06.PNG)

In this above snapshot see I have requested the data against asdasdas;*;*;* and since I have deleted the context and value associated with it am getting the result for default context as it exists. 


And when I come across the code I get to know that while requesting the data for any context it is checking inside the code based on the field "repositoryid" stored in property table in db.
And since we are only updating the context and value for the same key we always get the same "repoid".

This is what am talking about.

**Comment by edoarsla** [comment]
I think I know where the confusion is coming from.  Let's say you have a repository with only a single context level "Environment", and let's say you only defined a single context element for this level: "Development".

In this case, the UI will only allow you to:
1. Not specify a context
2. Specify "Development" context

However, when you're pulling context, you can specify anything for context you want to pull - for example, you could specify "Production", or "Foo", or "Development".

You will not get the error on any of them, as the resolution engine considers them all equally valid.  However, the configuration you get back for "Production" will be different then what you get back for "Development" - because they resolve different configuration.

Deletion of a context element really only matters for the UI, and for when you're POSTing configuration using the API.

**Comment by ShobhitGoel95** [comment]
Totally agree with that.
Let's make it simpler suppose I have a context as -> edo;shobhit;*;* and the value against this context is texting and now I have deleted the context then what should I expect in return?

**Comment by edoarsla** [comment]
So, if you have a property like this:
`{
   "key": "texting",
   "context": "edo;shobhit;;"
}`

the question is what will will happen if you delete the context - basically making this property now look like this:
`{
   "key": "texting",
   "context": ";;;"
}`

correct?

because this is a fully un-defined context, any request will match it.

**Comment by ShobhitGoel95** [comment]
Umm  So it means it'll give the output for the default one i.e., "*;*;*;*" or ";;;"

But if I have 3 contexts like this :

1. { "key": "someKey", "value": "texting", "context": "edo;shobhit;;" } 
2. { "key": "someKey", "value": "texting_with-one_context", "context": "edo;;;" } 
3. { "key": "someKey", "value": "texting_with_no_context", "context": ";;;" }
Now I have deleted the first context and in this case it is giving me the value for 2nd and 3rd context.'
even though am requesting the data against "context": "edo;shobhit;;"

**Comment by edoarsla** [comment]
I there's some terminology confusion here, so let's get out of the way first:
1. Property - has a key, value and a context.  I.e. `{ "key": "someKey", "value": "texting", "context": "edo;shobhit;;" }`
2. Context - specifies under which circumstance a property should be returned.  There are 2 "types" of context.  One, property context, which is assigned to the property (1.), and second a request context, which is the context you use to get properties that match.

When you say `But if I have 3 contexts like this :`, I read that as if you said: `there are 3 properties defined like this`.

When you say `Now I have deleted the first context`, I read that as if you said `I have deleted the first property`.

Now, if we're in agreement with terminology, and the first property is deleted, if in the UI you specified a context `"edo;shobhit;;` it is correct that both properties will be shown.  The reason is that both these properties are "resolved", or their contexts match your context request.  Keep in mind, this is only the case in the UI editor.

However, if you specified a "full context" (a context in which each element of the context has been specified with 1 value), like you would in the API call by your application, then, depending on your request context you would get only a single value for each key - where values match the context.  Look at http://docs.confighub.com/en/latest/pages/primer.html#context-resolution

The UI, or the editor, allows you to view configuration across multiple dimensions of configuration, whereas application requesting configuration via API call, HAS TO specify a full-context.  Due to the example of non-fully-quelified context you used as an example, I know you are viewing this in the UI.

To see the difference, or to see what your application would receive, try comparing that with a context like 'edo;shobhit;default`.
